client/helpers/FiniteStateMachine.js

```diff
@@ -0,0 +1,54 @@
+class FiniteStateMachine {
```

Comment:
FSM has strict state changes, it means that in order to change state there should be defined an action(s) which are allowed to make a state change
https://en.wikipedia.org/wiki/Finite-state_machine

also I think FSM should be combined and handled in scope of redux